wrapanapi/lenovo.py

```diff
@@ -1,42 +1,309 @@
-
 Used to communicate with providers without using CFME facilities
 """
+import json
+
+import requests
+from requests.exceptions import Timeout
+
 from base import WrapanapiAPIBase
 
 
 class LenovoSystem(WrapanapiAPIBase):
     """Client to Lenovo API
-
     Args:
         hostname: The hostname of the system.
         username: The username to connect with.
         password: The password to connect with.
-
     """
     _api = None
 
     _stats_available = {
-        'num_server': lambda self: len(self.list_server()),
+        'num_server': lambda self: len(self.list_servers()),
+        # 'num_server_with_host': lambda self: len(self.list_servers_with_host()),
     }
+    POWERED_ON = 8
+    POWERED_OFF = 5
+    STANDBY = 18
+    HEALTH_VALID = ("normal", "non-critical")
+    HEALTH_WARNING = ("warning")
+    HEALTH_CRITICAL = ("critical", "minor-failure", "major-failure", "non-recoverable", "fatal")
 
-    def __init__(self, hostname, username, password, **kwargs):
-        super(LenovoSystem, self).__init__(kwargs)
-        self.hostname = hostname
-        self.username = username
-        self.password = password
+    def __init__(self, hostname, username, password, protocol="https", port=443, **kwargs):
+        self.auth = (username, password)
+        self.url = '{}://{}:{}/'.format(protocol, hostname, port)
+        self._servers_list = None
+        self.kwargs = kwargs
 
     def __del__(self):
         """Disconnect from the API when the object is deleted"""
         # This isn't the best place for this, but this class doesn't know when it is no longer in
         # use, and we need to do some sort of disconnect based on the pyVmomi documentation.
-        raise NotImplementedError
+
+    def _service_instance(self, path):
+        """An instance of the service"""
+        try:
+            response = requests.get(self.url + path, auth=self.auth, verify=False)
+            return json.loads(response.content)
+        except Timeout:
+            return None
+
+    def _service_put(self, path, request):
+        """An instance of the service"""
+        try:
+            response = requests.put(self.url + path, data=json.dumps(request), auth=self.auth,
+                    verify=False)
+            return response
+        except Timeout:
+            return None
 
     @property
     def version(self):
         """The product version"""
-        raise NotImplementedError
+        response = self._service_instance("aicc")
+        return response['appliance']['version']
+
+    def list_servers(self):
+        response = self._service_instance("cabinet?status=includestandalone")
+        cabinets = response['cabinetList'][0]
+        nodes_list = cabinets['nodeList']
+
+        inventory = [node['itemInventory'] for node in nodes_list] if len(nodes_list) > 0 else []
+
+        nodes_from_chassis = []
+        if len(cabinets['chassisList']) > 0:
+            chassis_list = cabinets['chassisList'][0]
```

Comment:
**question** the structure/extraction of the normal node list, and the chassis node list looks quite different, is that imposed by lenovo?

Reply:
Yes @RonnyPfannschmidt, We have two different lists with nodes. The first is a nodeList directly and the second is the nodes inside the inventory on chassis.